src/ADAL.PCL/AcquireDeviceCodeHandler.cs

```diff
@@ -0,0 +1,103 @@
+            this.authenticator = authenticator;
+            this.callState = AcquireTokenHandlerBase.CreateCallState(this.authenticator.CorrelationId);
+            this.clientKey = new ClientKey(clientId);
+            this.resource = resource;
+            this.extraQueryParameters = extraQueryParameters;
+        }
+        
+        private string CreateDeviceCodeRequestUriString()
+        {
+            var deviceCodeRequestParameters = new DictionaryRequestParameters(this.resource, this.clientKey);
+
+            if (this.callState != null && this.callState.CorrelationId != Guid.Empty)
+            {
+                deviceCodeRequestParameters[OAuthParameter.CorrelationId] = this.callState.CorrelationId.ToString();
+            }
+            
+            if (PlatformPlugin.HttpClientFactory.AddAdditionalHeaders)
+            {
+                IDictionary<string, string> adalIdParameters = AdalIdHelper.GetAdalIdParameters();
+                foreach (KeyValuePair<string, string> kvp in adalIdParameters)
+                {
+                    deviceCodeRequestParameters[kvp.Key] = kvp.Value;
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(extraQueryParameters))
+            {
+                // Checks for extraQueryParameters duplicating standard parameters
+                Dictionary<string, string> kvps = EncodingHelper.ParseKeyValueList(extraQueryParameters, '&', false, this.callState);
+                foreach (KeyValuePair<string, string> kvp in kvps)
+                {
+                    if (deviceCodeRequestParameters.ContainsKey(kvp.Key))
+                    {
+                        throw new AdalException(AdalError.DuplicateQueryParameter, string.Format(AdalErrorMessage.DuplicateQueryParameterTemplate, kvp.Key));
+                    }
+                }
+
+                deviceCodeRequestParameters.ExtraQueryParameter = extraQueryParameters;
+            }
+
+            return new Uri(new Uri(this.authenticator.DeviceCodeUri), "?" + deviceCodeRequestParameters).AbsoluteUri;
+        }
+
+        internal async Task<DeviceCodeResult> RunHandlerAsync()
+        {
+            await this.authenticator.UpdateFromTemplateAsync(this.callState);
+            this.ValidateAuthorityType();
+            AdalHttpClient client = new AdalHttpClient(CreateDeviceCodeRequestUriString(), this.callState);
+            DeviceCodeResponse response = await client.GetResponseAsync<DeviceCodeResponse>(ClientMetricsEndpointType.DeviceCode);
+
+            if (!string.IsNullOrEmpty(response.Error))
+            {
+                throw new AdalException(response.Error, response.ErrorDescription);
+            }
+
+            return response.GetResult(clientKey.ClientId, resource);
+        }
+
+        private void ValidateAuthorityType()
+        {
+            if (this.authenticator.AuthorityType == AuthorityType.ADFS)
+            {
+                throw new AdalException(AdalError.InvalidAuthorityType,
```

Comment:
so, are you just throwing exception if authority type for acquiring user code is ADFS?